### PR TITLE
Fix Issue #34: Add new method Alefeld, Potra, Shi (1995)

### PR DESCRIPTION
**Pull request by huiyuxie**
### Summary

This pull request addresses issue #34 by adding a new method Alefeld to the SimpleNonlinearSolve.

### Changes

- Implemented the new method Alefeld(), a subtype of AbstractBracketingAlgorithm, in alefeld.jl
- Included alefeld.jl in SimpleNonlinearSolve.jl and exported Alefeld
- Added simple scalar function test like other bracketing methods (Falsi, Ridder, Brent) in basictests.jl

### Remaining Bugs

When tried to run runtests.jl, the terminal reported _''UndefVarError: Alefeld not defined''_ checked every detail about this repository but still could not fix it. 
Also, tried @ChrisRackauckas 's suggestion `]dev SimpleNonlinearSolve` and `]st` the terminal showed _''Status `~/.julia/environments/JuliaDev/Project.toml`
  [6e4b80f9] BenchmarkTools v1.3.2
  [2b5f629d] DiffEqBase v6.122.1
  [f6369f11] ForwardDiff v0.10.35
  [872c559c] NNlib v0.8.19
  [1bc83da4] SafeTestsets v0.0.1
  [727e6d20] SimpleNonlinearSolve v0.1.14 `~/.julia/dev/SimpleNonlinearSolve`
  [90137ffa] StaticArrays v1.5.19''_

### Next Steps

Add more complicated test to Alefeld() in basictests.jl and optimize the algorithm implementation. 

### Feedback

Any additional suggestions to help resolve the remaining bugs would be highly appreciated. Please let me know if there is anything I can do to improve the code or address the remaining bugs.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -5,22 +5,20 @@ version = "0.1.14"
...
 FiniteDiff = "6a86dc24-6348-571c-b903-95158fe2bd41"
 ForwardDiff = "f6369f11-7733-5829-9624-2563aa707210"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+NNlib = "872c559c-99b0-510c-b3b7-b6c96a88d5cd"
 Reexport = "189a3867-3050-52da-a836-e630ba90ab69"
 Requires = "ae029012-a4dd-5104-9daa-d747884805df"
+SafeTestsets = "1bc83da4-3b8d-516f-aca4-4fe02f6d838f"
 SciMLBase = "0bca4576-84f4-4d90-8ffe-ffa030f20462"
 SnoopPrecompile = "66db9d55-30c0-4569-8b51-7e840670fc0c"
+StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
```
Comment (ChrisRackauckas) [inline Project.toml:19]: These shouldn't be added as deps. They are only test deps.

File: src/alefeld.jl [hunk 2]
```diff
@@ -0,0 +1,165 @@
+"""
+`Alefeld()` 
+
+An implementation of algorithm 4.2 from [Alefeld](https://dl.acm.org/doi/10.1145/210089.210111).
+
+The paper brought up two new algorithms. Here choose to implement algorithm 4.2 rather than 
+algorithm 4.1 because, in certain sense, the second algorithm(4.2) is an optimal procedure.
+"""
+
+struct Alefeld <: AbstractBracketingAlgorithm end
```
Comment (ChrisRackauckas) [inline src/alefeld.jl:10]:
```suggestion
"""
struct Alefeld <: AbstractBracketingAlgorithm end
```

File: src/alefeld.jl [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        elseif f(b) * f(c) < 0
+            ā, b̄, d = c, b, a
+        end
+    end
+
+    return ā, b̄, d 
+end
+
+# Define subrotine function newton quadratic, return the approximation of zero
+function _newton_quadratic(f::Function, a, b, d, k)
```
Comment (ChrisRackauckas) [inline src/alefeld.jl:132]:
```suggestion
function _bracket(f::F, a, b, c) where F
    if iszero(f(c))
        ā, b̄, d = a, b, c
    else
        if f(a) * f(c) < 0 
            ā, b̄, d = a, c, b
        elseif f(b) * f(c) < 0
            ā, b̄, d = c, b, a
        end
    end

    return ā, b̄, d 
end

# Define subrotine function newton quadratic, return the approximation of zero
function _newton_quadratic(f::F, a, b, d, k) where F
```

Force specialization

File: src/alefeld.jl [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    for i in 1:k
+        rᵢ = rᵢ₋₁ - B * rᵢ₋₁ / (B + A * (2 * rᵢ₋₁ - a - b))
+        rᵢ₋₁ = rᵢ
+    end
+
+    return rᵢ₋₁
+end
+
+# Define subrotine function ipzero, also return the approximation of zero
+function _ipzero(f::Function, a, b, c, d)
```
Comment (ChrisRackauckas) [inline src/alefeld.jl:153]:
```suggestion
function _ipzero(f::F, a, b, c, d) where F
```

**Comment by huiyuxie** [comment]
All the reviews have been resolved, but the same problem persists. Here is the new package status:
_(@JuliaDev) pkg> st
      Status `~/.julia/environments/JuliaDev/Project.toml`
  [4fba245c] ArrayInterface v7.4.1
  [6e4b80f9] BenchmarkTools v1.3.2
  [2b5f629d] DiffEqBase v6.122.1
  [6a86dc24] FiniteDiff v2.19.0
  [f6369f11] ForwardDiff v0.10.35
  [872c559c] NNlib v0.8.19
  [189a3867] Reexport v1.2.2
  [ae029012] Requires v1.3.0
  [1bc83da4] SafeTestsets v0.0.1
  [0bca4576] SciMLBase v1.91.3
  [727e6d20] SimpleNonlinearSolve v0.1.14 `~/.julia/dev/SimpleNonlinearSolve`
  [66db9d55] SnoopPrecompile v1.0.3
  [90137ffa] StaticArrays v1.5.19
  [1e83bf80] StaticArraysCore v1.4.0_

File: src/alefeld.jl [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+`Alefeld()` 
+
+An implementation of algorithm 4.2 from [Alefeld](https://dl.acm.org/doi/10.1145/210089.210111).
+
+The paper brought up two new algorithms. Here choose to implement algorithm 4.2 rather than 
+algorithm 4.1 because, in certain sense, the second algorithm(4.2) is an optimal procedure.
+"""
+struct Alefeld <: AbstractBracketingAlgorithm end
+
+function SciMLBase.solve(prob::NonlinearProblem,
```
Comment (ChrisRackauckas) [inline src/alefeld.jl:11]:
```suggestion
function SciMLBase.solve(prob::IntervalNonlinearProblem,
```

**Comment by huiyuxie** [comment]
I suspect that there might be something wrong with a small step in this paper. I debugged the failed cases and found that the errors came from Newton-Quadratic, one of the subroutines in this paper. 

In Algorithm 4.2, Newton-Quadratic provides an interval [a, b] and an approximation c to bracket to approximate c again, which should ensure that c is within the interval (a, b). However, in some test cases, Newton-Quadratic returns c outside the interval [a, b]. I further examined this subroutine function and found that on page 330, equation (4), the induction is incorrect.
Screenshot 2023-03-29 at 8 36 23 PM
I could not get Br_{i-1} in the second line from P(r_{i-1}) in the first line. According to the definition of P(x), P(r_{i-1}) should be f(a)+B(r_{i-1}-a) + A(r_{i-1}-a)(r_{i-1}-b). This equation even does not hold for base case where r_{0} equals a or b.

However, this mistake seems so straightforward and wired to me that I am unsure whether it is an actual error or if I have misunderstood something. I would appreciate it if someone @ranocha could help me double-check this equation. Thank you!

**Comment by huiyuxie** [comment]
If the induction is indeed incorrect, I will directly use the formula in the first line of equation (4) rather than the second line.

**Comment by ChrisRackauckas** [comment]
https://github.com/SciML/SimpleNonlinearSolve.jl/actions/runs/4559629958/jobs/8051444496?pr=56#step:6:20534

Looks like the errors are all due to a typo?

**Comment by ChrisRackauckas** [comment]
I think you are correct. At first I was wondering if something factored out and canceled but no from the looks of the denominator that's not the case. Just looks like there was a typo.

**Comment by huiyuxie** [comment]
Thank you so much for your help @ChrisRackauckas! 

Here I have fixed two bugs in alefeld.jl:
1. As I mentioned earlier, there seem to be some typos in the _Newton-Quadratic_ function induction from the paper. I have now directly used the formula from the first line of equation (4) on page 330.
2. The error _"a\bar not defined"_ arose from the floating-point limit during the computation process, so I added a section to check whether c is equal to either a or b and constructed the solution with _"retcode = ReturnCode.FloatingPointLimit"_.

Now the majority of test cases have passed, but some test cases still fail. I believe there are still some corner cases that the code cannot handle very well. I am in the process of further debugging.
Screenshot 2023-03-30 at 2 06 47 PM

**Comment by ChrisRackauckas** [comment]
Tests seem to have passed. Looks good!
